pyvisa_py/protocols/hislip.py

```diff
@@ -0,0 +1,714 @@
+        self._sync.connect((ip_addr, port))
+        self._sync.settimeout(timeout)
+        self._sync.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        init = self.initialize()
+        if init.overlap != 0:
+            print("**** prefer overlap = %d" % init.overlap)
+
+        # open the asynchronous socket and send an initialize packet
+        self._async = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._async.connect((ip_addr, port))
+        self._async.settimeout(timeout)
+        self._async.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        self._async_init = self.async_initialize(session_id=init.session_id)
+
+        # get the maximum message size
+        max_msg_size = self.async_maximum_message_size(1 << 24)
+        # print("HEADER_SIZE = %s, async_maximum_message_size = %s" % (HEADER_SIZE, max_msg_size))
+        # print("max_payload_size = %s" % (max_msg_size - HEADER_SIZE))
+        self.max_payload_size = max_msg_size - HEADER_SIZE
+
+        self.timeout = 10
+        self.rmt = 0
+        self.receiver = None
+        self.message_id = 0xFFFFFF00
+
+    # ================ #
+    # MEMBER FUNCTIONS #
+    # ================ #
+
+    def close(self):
+        self._sync.close()
+        self._async.close()
+
+    @property
+    def timeout(self):
+        """returns the timeout value in seconds for both the sync and async sockets"""
+        return self._timeout
+
+    @timeout.setter
+    def timeout(self, val):
+        """sets the timeout value in seconds for both the sync and async sockets"""
+        self._timeout = val
+        self._sync.settimeout(self._timeout)
+        self._async.settimeout(self._timeout)
+
+    def send(self, data):
+        """Sends the data on the synchronous channel.  More than one packet
+        may be necessary in order to not exceed max_payload_size."""
+        data_view = memoryview(data)
+        num_bytes_to_send = len(data)
+
+        # send the data in chunks of self.max_payload_size bytes at a time
+        while num_bytes_to_send > 0:
+            if num_bytes_to_send <= self.max_payload_size:
+                assert len(data_view) == num_bytes_to_send
+                self.send_data_end_packet(data_view)
+                bytes_sent = num_bytes_to_send
+            else:
+                self.send_data_packet(data_view[: self.max_payload_size])
+                bytes_sent = self.max_payload_size
+
+            data_view = data_view[bytes_sent:]
+            num_bytes_to_send -= bytes_sent
+
+        return len(data)
+
+    def receive(self, max_len=4096):
+        """
+        receive data on the synchronous channel, terminating after
+        max_len bytes or after receiving a DataEnd message
+        """
+
+        # if we don't already have a Receiver object, create one
+        if self.receiver is None:
+            self.receiver = Receiver(self._sync, self.last_message_id)
```

Comment:
Is there an advantage to delaying the receiver creation in this fashion. Could it not be created as part of the initialization ?

Reply:
This needs some work.  Let me take a look and play with a couple of ideas.